### DISCUSSION TITLE
Join to or leave room by "external" process

### DISCUSSION THREAD

**Discussion by bokolob** [miguelgrinberg/python-socketio#724] [Q&A]
Hello, I can't understand clearly if  I can join some user to some room on the server that he isn't connected to. As I can get from code - it doesn't work because leave and join  both work with local dicts, that are not shared  between machines.

It seems that join and leave calls could work only when are called on the server (even different gunicorn process on the same server!) where client is connected.  Moreover, user can stay joined some room on one server which he had left on another (not for a long time probably).

For example - I use socketio to push some messages to users, and If user is joined to some room means that he has rights to particular kind of messages.  Now, administrator can subscribe this user to particular room, to give him some rights, but he is connected to different server.

I believe this calls should be processed in _thread as well as close_room is.

**Comment by miguelgrinberg** [comment]
You are correct, joining and leaving rooms must be called on the server the target user is connected to.

**Comment by bokolob** [comment]
What if I fix that?

**Comment by miguelgrinberg** [comment]
Well, it is not really a bug.

The most common case is that a client makes a request to join a room, so there is really no way this request can end up in the wrong server. The case that is not covered is a when an action performed by one client or by the server on its own requires some other client to be added to a room. Is that something that you need?

**Comment by bokolob** [comment]
Yes. Not only added but removed too.

**Comment by bokolob** [comment]
But, it seems to be essential, even in chat applications users can be kicked by moderators..

**Comment by miguelgrinberg** [comment]
Okay. If you want to give it a go I have no problem adding it.

**Comment by Kub-AT** [comment]
Is there any workaround to join/leave room from external process ?

**Reply by miguelgrinberg** [reply]
This has not been implemented, but I stand by my offer above, if you want to add it I'll be happy to consider it for inclusion in the library.
